streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java

```diff
@@ -53,34 +85,121 @@ public void process(final K key, final Change<V> value) {
     }
 
-    private boolean shouldForward(final Change<V> value) {
-        return !(value.newValue == null && suppress.suppressTombstones());
+    @SuppressWarnings("unchecked")
+    private <T> Serde<T> castSerde(final Serde<?> untyped, final String keyOrValue) {
+        try {
+            return (Serde<T>) untyped;
+        } catch (final ClassCastException e) {
+            throw new TopologyException(
+                "Unable to use the configured default " + keyOrValue +
+                    " serde. Ensure that the correct " + keyOrValue + " serde is set upstream of " + myNode + ".",
+                e
+            );
+        }
     }
 
     private long definedRecordTime(final K key) {
         return suppress.getTimeDefinition().time(internalProcessorContext, key);
     }
 
+    private void buffer(final K key, final Change<V> value) {
+        final long time = suppress.getTimeDefinition().time(internalProcessorContext, key);
+        final ProcessorRecordContext recordContext = internalProcessorContext.recordContext();
+
+        buffer.put(time, key, new ContextualRecord<>(time, value, recordContext), computeRecordSize(key, value, recordContext));
+    }
+
+    private void evictOldEnoughRecords(final long streamTime) {
+        System.out.println("st: " + streamTime);
+        final Iterator<KeyValue<TimeKey<K>, ContextualRecord<Change<V>>>> iterator = buffer.iterator();
+        while (iterator.hasNext()) {
+            final KeyValue<TimeKey<K>, ContextualRecord<Change<V>>> next = iterator.next();
+            final long bufferTime = next.key.time();
+            if (bufferTime <= streamTime - suppressDuration.toMillis()) {
+                final K key = next.key.key();
+                final ContextualRecord<Change<V>> contextualRecord = next.value;
+
+                setNodeAndForward(key, contextualRecord);
+                iterator.remove();
+            } else {
+                break;
+            }
+        }
+        System.out.println(buffer);
+    }
+
+    private void enforceSizeBound() {
+        if (overCapacity()) {
+            switch (suppress.getBufferConfig().bufferFullStrategy()) {
+                case EMIT:
+                    final Iterator<KeyValue<TimeKey<K>, ContextualRecord<Change<V>>>> iterator = buffer.iterator();
+                    while (overCapacity() && iterator.hasNext()) {
+                        final KeyValue<TimeKey<K>, ContextualRecord<Change<V>>> next = iterator.next();
+                        final K key = next.key.key();
+                        final ContextualRecord<Change<V>> contextualRecord = next.value;
+                        // to be safe, forward before we delete
+                        setNodeAndForward(key, contextualRecord);
+                        iterator.remove();
+                    }
+                    return;
+                case SHUT_DOWN:
+                    throw new RuntimeException("TODO: request graceful shutdown"); // TODO: request graceful shutdown
```

Comment:
Oh yeah, I was meaning to figure out the right exception to throw to achieve a nice shutdown (I think any runtime exception will do it, but is there a semantically best one?)